ext/flash/lib.rs

```diff
@@ -0,0 +1,1451 @@
+  if let Some(response) = maybe_body {
+    let _ = sock.write(format!("{:x}", response.len()).as_bytes());
+    let _ = sock.write(b"\r\n");
+    let _ = sock.write(&response);
+    let _ = sock.write(b"\r\n");
+  }
+
+  // server is done writing and request doesn't want to kept alive.
+  if shutdown && close {
+    match &mut sock.inner {
+      InnerStream::Tcp(stream) => {
+        // Typically shutdown shouldn't fail.
+        let _ = stream.shutdown(std::net::Shutdown::Both);
+      }
+      InnerStream::Tls(stream) => {
+        let _ = stream.sock.shutdown(std::net::Shutdown::Both);
+      }
+    }
+  }
+}
+
+#[op]
+fn op_flash_respond_chuncked(
+  op_state: &mut OpState,
+  server_id: u32,
+  token: u32,
+  response: Option<ZeroCopyBuf>,
+  shutdown: bool,
+) {
+  let flash_ctx = op_state.borrow_mut::<FlashContext>();
+  let ctx = flash_ctx.servers.get_mut(&server_id).unwrap();
+  match response {
+    Some(response) => {
+      respond_chunked(ctx, token, shutdown, Some(&response));
+    }
+    None => {
+      respond_chunked(ctx, token, shutdown, None);
+    }
+  }
+}
+
+#[op]
+async fn op_flash_write_resource(
+  op_state: Rc<RefCell<OpState>>,
+  response: StringOrBuffer,
+  server_id: u32,
+  token: u32,
+  resource_id: deno_core::ResourceId,
+) -> Result<(), AnyError> {
+  let resource = op_state.borrow_mut().resource_table.take_any(resource_id)?;
+  let sock = {
+    let op_state = &mut op_state.borrow_mut();
+    let flash_ctx = op_state.borrow_mut::<FlashContext>();
+    let ctx = flash_ctx.servers.get_mut(&server_id).unwrap();
+    ctx.response.remove(&token).unwrap().socket()
+  };
+
+  drop(op_state);
+  let _ = sock.write(&response);
+
+  #[cfg(unix)]
+  {
+    use std::os::unix::io::AsRawFd;
+    if let InnerStream::Tcp(stream_handle) = &sock.inner {
+      let stream_handle = stream_handle.as_raw_fd();
+      if let Some(fd) = resource.clone().backing_fd() {
+        // SAFETY: all-zero byte-pattern is a valid value for libc::stat.
+        let mut stat: libc::stat = unsafe { std::mem::zeroed() };
+        // SAFETY: call to libc::fstat.
+        if unsafe { libc::fstat(fd, &mut stat) } >= 0 {
+          let _ = sock.write(
+            format!("Content-Length: {}\r\n\r\n", stat.st_size).as_bytes(),
+          );
+          let tx = sendfile::SendFile {
+            io: (fd, stream_handle),
+            written: 0,
+          };
+          tx.await?;
+          return Ok(());
+        }
+      }
+    }
+  }
+
+  let _ = sock.write(b"Transfer-Encoding: chunked\r\n\r\n");
+  loop {
+    let vec = vec![0u8; 64 * 1024]; // 64KB
+    let buf = ZeroCopyBuf::new_temp(vec);
+    let (nread, buf) = resource.clone().read_return(buf).await?;
+    if nread == 0 {
+      let _ = sock.write(b"0\r\n\r\n");
+      break;
+    }
+    let response = &buf[..nread];
+
+    let _ = sock.write(format!("{:x}", response.len()).as_bytes());
+    let _ = sock.write(b"\r\n");
+    let _ = sock.write(response);
+    let _ = sock.write(b"\r\n");
```

Comment:
are these buffered at all? are these all separate syscalls? Presumably the return value is the number of bytes written - it's possible that the complete string isn't written?